test/test_autograd.py

```diff
@@ -5421,6 +5421,26 @@ def get_refs(with_backward):
         gc.collect()
         self.assertTrue(ref.expired())
 
+    def test_no_unnecessary_unwrapping(self):
+        a = torch.ones(5, requires_grad=True)
+        b = a * a
+        c = a * b
+
+        # a is leaf
+        self.assertIs(b.grad_fn._saved_self, a)
+        self.assertIs(b.grad_fn._saved_other, a)
+        self.assertIs(c.grad_fn._saved_self, a)
+
+        # b is not an output
+        self.assertIs(c.grad_fn._saved_other, b)
+
+        c.sum().backward()
+
+        with self.assertRaisesRegex(RuntimeError, "after they have already been freed"):
+            c.grad_fn._saved_self
+
+        self.assertEqual(a, torch.ones(5))
```

Comment:
Why do you check the value of a here? Did you meant to check the value of a.grad? If so, maybe use another value than 1 for a so that you get non-1 gradients.

Reply:
I meant to check that the value of `a` has not been cleared by `release_variables`.

Reply:
Ah! Maybe leave a comment there "a is still untouched"?
Also, unless we do something very bad with refcounting, there is no reason for a to have disapeared here.